web/src/composebox_typeahead.ts

```diff
@@ -42,41 +52,144 @@ import {user_settings} from "./user_settings";
 // highlighter that escapes (i.e. one that calls
 // typeahead_helper.highlight_with_escaping).
 
+// ---------------- TYPE DECLARATIONS ----------------
+// There are many types of suggestions that can show
+// up in the composebox typeahead, but they are never
+// mixed together. So a user can see a list of stream
+// suggestions in one situation, and a list of user
+// suggestions in another, but never both at the same
+// time.
+// We use types with a "type" field to keep track
+// of and differentiate each kind of suggestion,
+// because we handle multiple kinds of suggestions
+// within shared code blocks.
+type SlashCommand = {
+    text: string;
+    name: string;
+    aliases: NamedCurve;
+    placeholder?: string;
+};
+export type SlashCommandSuggestion = SlashCommand & {type: "slash"};
+
+export type LanguageSuggestion = {
+    language: string;
+    type: "syntax";
+};
+
+type TopicSuggestion = {
+    topic: string;
+    type: "topic_list";
+};
+
+type TimeJumpSuggestion = {
+    message: string;
+    type: "time_jump";
+};
+
+type TopicJumpSuggestion = {
+    message: string;
+    type: "topic_jump";
+};
+
+type MentionList = {
+    type: "mention";
+    is_silent: boolean;
+    list: (UserGroupPillData | UserOrMentionPillData)[];
+};
+
+type TypeaheadSuggestion =
+    | UserGroupPillData
+    | UserOrMentionPillData
+    | StreamPillData
+    | TopicJumpSuggestion
+    | TimeJumpSuggestion
+    | LanguageSuggestion
+    | TopicSuggestion
+    | EmojiSuggestion
+    | SlashCommandSuggestion;
+
+// We also often deal with a list of suggestions (which have
+// to be the same type), and we add a "type" field to that as
+// well so that we don't have to rely on a more hacky checking
+// what type of suggestion is contained in the list (which could
+// also be an empty list).
```

Comment:
So both the `Suggestion` types and `List` types have a `type` field, and this is because we often work with the lists and I want to be able to easily get what kind of list it is. But when an item is selected in the typeahead we just have the individual item and we need to know the type from there too. This feels a little janky but I couldn't think of something better.